Providers/TopSitesProvider.swift

```diff
@@ -0,0 +1,144 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0
+
+import Foundation
+import Shared
+import UIKit
+import Storage
+import SyncTelemetry
+
+/// A provider for frecency and pinned top sites, used for the home page and widgets
+protocol TopSitesProvider {
+
+    /// Get top sites from frecency and pinned tiles
+    func getTopSites(numberOfMaxItems: Int,
+                     completion: @escaping ([Site]?) -> Void)
+
+    /// Fetches the default top sites
+    func defaultTopSites(_ prefs: Prefs) -> [Site]
+
+    /// Default maximum number of items fetched for frecency
+    static var numberOfMaxItems: Int { get }
+
+    /// Hide with search param is defined by adMarketplace, indicates this URL was registered through sponsored clicks
+    /// and should not show in top sites
+    var hideWithSearchParam: String { get }
+
+    /// Default key for suggested sites
+    var defaultSuggestedSitesKey: String { get }
+}
+
+extension TopSitesProvider {
+
+    func getTopSites(numberOfMaxItems: Int = Self.numberOfMaxItems,
+                     completion: @escaping ([Site]?) -> Void) {
+        getTopSites(numberOfMaxItems: numberOfMaxItems, completion: completion)
```

Comment:
Why is this calling itself? This looks like a never-ending recursion to me.

Reply:
We can't have default params in protocol, that's one way of doing it. With that we can call the protocol function with only `getTopSites(completion:)` instead of defining `numberOfMaxItems`. It's not doing a recursion in that case.